### PR TITLE
(#3268) Build NCIDS Header Block

### LINKED ISSUE #3268: Story: Build a NCIDS Header Block in Drupal
## As a Drupal Site Builder, I want to be able to add a NCIDS Header to my Drupal site as a configuration block, so that I can provide my users a way to navigate the site following the NCIDS without having to enter complicated HTML

> **ESTIMATE** 5

### Requirements:
* Set up new NCIDS Header Block
  * Additional fields:
    * Home Page Title - aria label used in place of the alt tag for the logo
    * Search Results Page - relative URL to use as the search results page
    * Autosuggest Collection: freeform text field for search autosuggest collection
      * NOTE: If there is no autosuggest, leave this empty, but also do not wire up the autosuggest in the search box component 
* We would like the logo to be managed within the Theme settings and not requiring a release. We also need to support alternate languages and other text in the image as settings as well.
   * Use page langcode as SVG indicator thingy (#)
   * Template will have to output the langcode
   * Figure out how to allow SVGs in Drupal
* Analytics for banner, toggle, search box, and primary nav

### Notes:
* There should be a [cache-context added to the cache information](https://www.drupal.org/docs/drupal-apis/cache-api/cache-contexts#core-contexts) on the block that will vary it based on the nav that is being shown.
   * IDK what this might be at this point as the contexts are from the request standpoint, NOT from the page content standpoint.
   * Basically http://www.devbox/ and http://www.devbox/api-test should show different menus on your machine and on ODEs
* This will render the static elements of the `nci-extended` header:
   * The logo, and any alternate text
   * The primary navigation
       * If says "Hide in Main Nav", do not show - default to checked
     * Navigation label over section label
     * If the primary nav item does not have a megamenu, don't output `data-menu-id` and make it a link to the page
   * The search box (with autosuggest configured)
      * This requires the following information:
         * The URL of the search results page
         * The API base url. This should be available via Javascript as `window.CDEConfig.sitewideSearchConfig.searchApiServer`
            * So however you initialize that autosuggest instance, use that! 
               * The text input for the search box should have a `data-autosuggest-collection` attribute that contains the configured collection. 
            * And no, you should not be asking "how does Drupal read the value from the Javascript"
   * Any other bits and pieces. (like the mobile menu button)

This will NOT include any of the Javascript elements defined in the tickets below. They should enhance the page on load and can be handled separately.
* NCIOCPL/cgov-digital-platform#3277
* NCIOCPL/cgov-digital-platform#3278
* NCIOCPL/cgov-digital-platform#3279

### Acceptance Criteria

```gherkin
Scenario: A site admin sets the logo and alternate text in the Drupal ???
  Given User is a Site Admin
    And [some more context]
  When  Drupal renders the header
  Then  The logo appears as a link with the "correct accessibility information"
    And [another outcome]
```

* Site admin adds a ncids header sets "The accessible label for the home page link."
  * header renders with the following
     ```
     <a href="/" aria-label="<THE LABEL>">
        <svg role="img" width="385" height="38" aria-hidden="true">
          <use href="/profiles/custom/cgov_site/themes/custom/ncids_trans/logo.svg#en"></use>
        </svg>
      </a>
     ```
* Site admin adds a ncids header without "Search Results Page"
   * No search box appears
* Site admin selects a page from the system for the "Search Results Page"
   * when the user submits the search form they are taken to the search results page with their keyword
* Site admin sets the "The collection name for the autosuggest API."
   * the input has a data-autosuggest-collection attribute with that value
* Site admin goes to /admin/appearance/settings/ncids_trans and expands the LOGO IMAGE container. Unchecks the checkbox. Clicks choose file and uploads a different logo.
   * The new logo appears correctly for English and Spanish
* Site admin edits a Cgov Site Section and checks the "Set Main Nav Root" checkbox
   * All the immediate child Cgov Site Sections will display in the primary navigation if the following conditions are met:
      * The child site section does not have "Hide in main nav" checked under the "Navigation Display Options" heading
      * The child section has a landing page
      * The landing page has been published 
   * 
 
---

### Prerequisites
- [ ] https://github.com/NCIOCPL/ncids/wiki/Technical-NCI-Header-Design

### LINKED ISSUE #3277: Build: Redesign megamenu JS
### Build the JS component of the redesign megamenu based on design specs/requirements documented in this ticket for pulling data into the menu

> **ESTIMATE** 8

This story is to create a CgdpMegaMenuAdapter for use in initializing the NCIExtendedHeaderWithMegaMenu component. This will use the API that is being created in #3265.  Below is a rough sequence diagram of how this will work.

NOTE: If a primary navigation item does NOT have megamenu content, then clicking on that item should go directly to that section and not open up an empty mega menu. This applies to Spanish Grants and Training specifically.

```mermaid
sequenceDiagram
    actor Editor
    actor Visitor
    Editor->>/taxonomy/term/32/edit: Enters YAML Content for Mega Menu in Drupal
    Drupal->>/taxonomy/term/32/edit: Validates YAML
    Drupal->>/taxonomy/term/32/edit: Stores YAML

    Visitor->>Web Page: load
    Web Page->>+NCIExtendedHeaderWithMegaMenu: create()
    Visitor->>Web Page: click About Cancer (id: 32)
    Web Page->>NCIExtendedHeaderWithMegaMenu: handle click event
    NCIExtendedHeaderWithMegaMenu->>cgdpMegaMenuAdapter: getMegaMenuContent(id: 32)
    cgdpMegaMenuAdapter-)/taxonomy/term/32/mega-menu: Fetch Data
    /taxonomy/term/32/mega-menu-->>cgdpMegaMenuAdapter: json data
    cgdpMegaMenuAdapter-->>cgdpMegaMenuAdapter: Parse json and create HTML
    cgdpMegaMenuAdapter->>NCIExtendedHeaderWithMegaMenu: HtmlElement
    NCIExtendedHeaderWithMegaMenu-->>Web Page: Open Mega Menu Overlay
    NCIExtendedHeaderWithMegaMenu-->>Web Page: Display Html from adapter
  
```
The ID passed into getMegamenuContent will come from the data-menu-id attribute on the navigation item.
Each navigation item in the primary nav should have a `data-menu-id` attribute with the ID of that Drupal site section.

### InVision: 
  * DSM link: *insert link here, or enter N/A if not needed*
  * Desktop mockup: *insert link here, or enter N/A if not needed*
  * Tablet mockup: *insert link here, or N/A if not needed*
  * Mobile mockup: *insert link here, or N/A if not needed*

### Acceptance Criteria for Behaviors 
*If a component changes between breakpoints, or things show/hide, that gets captured here* 

```gherkin
Scenario: User is activating the megamenu
  Given user is on c.gov on desktop
  When the user clicks on an L1 item in the main nav
  Then the megamenu for that L1 will appear below 
     And the L1 will be highlighted in the main nav
     And the caret for the L1 will flip to point upwards
  When the user hovers on any link in the megamenu
  Then the link will receive an underline
  When the user clicks on the L1 item that appears in the megamenu ("Explore L1")
  Then the landing page will load
     And the megamenu will close
     And an indicator will appear in the main nav under that L1
  When an L2 is clicked on 
  Then the page will load
     And the megamenu will close
     And an indicator will appear in the main nav under the L1 parent
  When an L3 is clicked on 
  Then the page will load
     And the megamenu will close
     And an indicator will appear in the main nav under the L1 parent
  When the user clicks on an L1 item in the main nav
  Then the megamenu for that L1 will appear below 
  When the user clicks outside of the megamenu overlay or on the L1 again
  Then the megamenu will close
  When the user clicks on an L1 item in the main nav
  Then the megamenu for that L1 will appear below 
  When another L1 is clicked on
  Then the megamenu for that L1 will appear below 
```

### Acceptance Criteria for Analytics
``` gherkin
Scenario:
Given a user is on any page of cancer.gov
When the user clicks on a primary nav link that opens the mega menu
Then there should be an NCIDataLayer Other event 
  And the event has the name “PrimaryNav:Open”
  And the event has the the linkname “PrimaryNav:Open”
  And this event should have the following data:
      | key          | value                                                          |
      | topLevelMenu | Top nav section menu of menu that is opened (ie: About Cancer) |
      | location     | Primary Nav                                                       |
```
``` gherkin
Scenario:
Given a user is on any page of cancer.gov
When the user clicks on a primary nav link that goes directly to the page (no mega menu)
Then there should be an NCIDataLayer Other event 
  And the event has the name “PrimaryNav:LinkClick”
  And the event has the linkname “PrimaryNav:LinkClick”
And have the following data:
    | key          | value                                                          |
      | location     | Primary Nav                                                     |     
      | primaryNavItem | Top nav section menu of menu that is opened (ie: About Cancer) |
      | listHeading  | This should pass the same value as the link text (the name of the button)               |
      | linkText     | Text of the link clicked                                       |
      | linkType | This would be "Primary Nav Button" for this type of click |
      | listHeadingNumber  | Should pass as 0 in this scenario.                  |
      | listItemNumber   | Should pass as 0 in this scenario.     |
```

``` gherkin
Given a user is on any page of cancer.gov
  And the mega menu has been opened
When the user clicks on a link in the mega menu
Then there should be an NCIDataLayer Other event 
  And the event has the name “PrimaryNav:LinkClick”
  And the event has the linkname “PrimaryNav:LinkClick”
And have the following data:
      | key          | value                                                          |
      | location     | Primary Nav                                                     |     
      | primaryNavItem | Top nav section menu of menu that is opened (ie: About Cancer) |
      | listHeading  | _Subheading of link clicked (ie: Advanced Cancer) . If link clicked is primary nav label (i.e. Explore About Cancer), this value should pass the linkText (i.e. Explore About Cancer)          |
      | linkText     | Text of the link clicked                                       |
      | linkType | This would be "List Heading", "List Item" or "Primary Nav Label" |
      | listHeadingNumber  | This is the number of the list heading within the larger list (ex: for About Cancer Menu, Causes & Prevention would be 2). If the user clicks on a primary nav label (ex: Explore About Cancer) this value should be set to 0.                  |
      | listItemNumber   | This is the number within a list heading of the links (ex: for About Cancer Menu, Cancer Statistics would be 2. If link clicked is a primary nav label or list heading this should be 0.     |
      
```
---

### Prerequisites
- *required dependency for identifying the problem, scope and then solution for this issue*

- [x] #3268
- [x] #3265
- [ ] https://github.com/NCIOCPL/ncids/issues/297

### Tasks
- [ ] Add `data-menu-id` to each of the items in the primary nav in the template

### LINKED ISSUE #3459: Enabler: Create a mechanism for installing dev-only blocks onto dev systems
# Create a mechanism for installing dev-only blocks onto dev systems
> **ESTIMATE** TBD

The NCIDS Header requires entity IDs for setting the search URL. This is because the search URL uses a lookup field to find the results page, which is nice for users. We don't want to/can't hardcode a node ID in the Profile's install configuration files, or else new ACSF sites would not work correctly. As such we need some sort of command like the YAML content commands to load these "dev" blocks with the ODE content. Blocks are a weird configuration thing where it is not content, but the UX is meant for site maintainers to add new blocks (configs) to different areas on the site with different options. So this should go WITH the yaml content.

### LINKED ISSUE #3474: Enabler: Add Support for Templating in JS for Complex Components
# Add Support for Templating in JS for Complex Components
> **ESTIMATE** TBD

The Mega Menu content is complicated HTML that needs to be generated in JS. Doing this through a bunch of appendChild would result in unreadable code. (I partially did it already and it is a mess) This needs to support adding click handlers, which rules out many of the templating engines. One possibility is to implement JSX. JSX does not require React, it only requires a function for handling element & fragment creation. Babel and Typescript both support a way of setting those functions. This will not be a full on implementation of React, and only support basic click handling.

### Related Resources
* https://betterprogramming.pub/how-to-use-jsx-without-react-21d23346e5dc
* https://blog.r0b.io/post/using-jsx-without-react/
* https://itnext.io/lessons-learned-using-jsx-without-react-bbddb6c28561

### PR DESCRIPTION
**Pull request by bryanpnih**
ODE: http://ncigovcdode481.prod.acquia-sites.com
Closes #3459 
Closes #3474 
Closes #3268
Closes #3277

### CODE REVIEW

**Comment by blilianyu** [comment]
A couple of issues:
- Mega Menu for About NCI “Explore About NCI >” the arrow is breaking on the 2nd line (in [PR527](https://github.com/NCIOCPL/ncids/pull/527) when I put in "Explore About NCI >" it was on one line)
About NCI Mega Menu

- Mobile menu isn’t loading from 1023 and less
Mobile Menu

- on Spanish, mega menu is cutting off “p” in “Subvenciones y capacitación
Spanish Mega Menu
"

**Comment by meglake** [comment]
@bryanpizzillo updates to @blilianyu list above per our discussion:

1. Confirmed this is an accepted issue and will be re-evaluated when post launch mega menu testing is conducted (this is recorded in our Redesign/DS Teams space under the 'Team Resources' wiki)
2. Confirmed this is a non-issue as mobile menu is not part of this PR review
3. Confirmed this will be an issue that @olitharp-nci will look into as part of her current sprint work.

File: docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/core/jsx/cgdp-jsx.ts [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+ * @param parent The parent
+ * @param child The thing to add (can be an array);
+ */
+const appendChild = (parent: Node, child: Node | string) => {
+	if (Array.isArray(child)) {
+		child.forEach((nestedChild) => appendChild(parent, nestedChild));
+	} else {
+		parent.appendChild(
+			typeof child === 'string' || child instanceof String
+				? document.createTextNode(child as string)
```
Comment (olitharp-nci) [inline docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/core/jsx/cgdp-jsx.ts:28]: You said don't just createTextNode :)

**Review by olitharp-nci — CHANGES_REQUESTED** [review]

File: docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/usa-sidenav/_usa-sidenav.scss [hunk 2]
```diff
@@ -1 +1 @@
-@import '@nciocpl/ncids-css/scss/components/usa-sidenav/nci-sidenav';
+@import '@nciocpl/ncids-css/scss/components/usa-sidenav/nci-sidenav'
```
Comment (olitharp-nci) [inline docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/usa-sidenav/_usa-sidenav.scss:1]:
```suggestion
@import '@nciocpl/ncids-css/scss/components/usa-sidenav/nci-sidenav';
```

Comment (bryanpizzillo) [inline docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/usa-sidenav/_usa-sidenav.scss:1]: Fix this

**Review by olitharp-nci — CHANGES_REQUESTED** [review]

File: docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/core/jsx/cgdp-jsx.ts (hunk 1 above)
Comment (bryanpizzillo) [reply docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/core/jsx/cgdp-jsx.ts:28]: Haha. JK. JK. It already does encode.

File: docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/nci-header/cgdp-mega-menu/components/cgdp-mega-menu.tsx [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+					<div className='nci-megamenu__items-pane grid-col-9'>
+						{mmItemsPaneContents}
+					</div>
+				</div>
+			</div>
+		</div>
+	);
+}
+
+const getCgdpMegaMenu = (mmData: CgdpMegaMenuData, id: string): HTMLElement => {
```
Comment (olitharp-nci) [inline docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/nci-header/cgdp-mega-menu/components/cgdp-mega-menu.tsx:90]: jsdoc here

File: docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/nci-header/cgdp-mega-menu/components/cgdp-mega-menu-list.tsx [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+import type { CgdpMegaMenuListData, CgdpMegaMenuListItemData } from '../types';
+
+type CgdpMegaMenuListProps = {
+	linkTracking: (data: {[key: string]: string|number}) => void;
+	list: CgdpMegaMenuListData;
+	row: number;
+	col: number;
+}
+
+const CgdpMegaMenuList: React.FunctionComponent<CgdpMegaMenuListProps> = (
```
Comment (olitharp-nci) [inline docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/nci-header/cgdp-mega-menu/components/cgdp-mega-menu-list.tsx:11]: jsdoc here

**Comment by kate-mashkina** [comment]
Verified a bunch of scenarios from #3268 - didn't find any gotchas
discussed YAML updates for the megaMenu content with @bryanpizzillo

File: docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/nci-header/cgdp-mega-menu/components/cgdp-mega-menu-list.tsx (hunk 4 above)
Comment (bryanpizzillo) [reply docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/nci-header/cgdp-mega-menu/components/cgdp-mega-menu-list.tsx:15]: I updated this.

File: docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/nci-header/cgdp-mega-menu/components/cgdp-mega-menu.tsx (hunk 3 above)
Comment (bryanpizzillo) [reply docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/nci-header/cgdp-mega-menu/components/cgdp-mega-menu.tsx:93]: Updated

**Comment by kate-mashkina** [comment]
primaryNavLinkClickHandler is missing the following:
linkType
listHeadingNumber 
listItemNumber 
listHeading

**Comment by kate-mashkina** [comment]
> primaryNavLinkClickHandler is missing the following: linkType listHeadingNumber listItemNumber listHeading

this is fixed

File: docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/nci-header/__tests__/nci-header.mega-menu.dom.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+					</form>
+				</div>
+			</div>
+
+			<nav aria-label="Primary navigation" class="nci-header-nav">
+				<div class="nci-header-nav__inner">
+					<ul class="nci-header-nav__primary">
+						<li class="nci-header-nav__primary-item">
+							<a data-menu-id="1234" class="nci-header-nav__primary-link" href="#">
+								First Section
```
Comment (olitharp-nci) [inline docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/nci-header/__tests__/nci-header.mega-menu.dom.ts:38]:
```suggestion
								<span>First Section</span>
```

File: docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/nci-header/__tests__/nci-header.mega-menu.dom.ts [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+							</a>
+						</li>
+						<li class="nci-header-nav__primary-item">
+							<a data-menu-id="2345" class="nci-header-nav__primary-link" href="#">
+								Second Section
```
Comment (olitharp-nci) [inline docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/nci-header/__tests__/nci-header.mega-menu.dom.ts:43]:
```suggestion
								<span>Second Section</span>
```

File: docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/nci-header/__tests__/nci-header.mega-menu.dom.ts [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+							</a>
+						</li>
+						<li class="nci-header-nav__primary-item">
+							<a class="nci-header-nav__primary-link" data-megamenu-disabled="true" href="#">
+								Third Section
```
Comment (olitharp-nci) [inline docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/nci-header/__tests__/nci-header.mega-menu.dom.ts:48]:
```suggestion
								<span>Third Section</span>
```

File: docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/nci-header/cgdp-mega-menu/__tests__/cgdp-megamenu-adapter.test.ts [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+		expect(consoleError).toHaveBeenNthCalledWith(
+			2,
+			'Could not fetch megamenu for 1234.'
+		);
+		scope.isDone();
+	});
+
+	// Click heading
+	// Click sublink
+	// Click Primary
```
Comment (olitharp-nci) [inline docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/nci-header/cgdp-mega-menu/__tests__/cgdp-megamenu-adapter.test.ts:120]: are these todos?

File: docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/nci-header/cgdp-mega-menu/components/cgdp-mega-menu.tsx [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+									headingUrl: mmData.primary_nav_url,
+									headingIndex: 0,
+									label: mmData.primary_nav_label,
+									url: mmData.primary_nav_url,
+									listItemNumber: 0
+								});
+							}}
+							className='nci-megamenu__primary-link'
+							href={mmData.primary_nav_url}
+						>{mmData.primary_nav_label}</a>
```
Comment (olitharp-nci) [inline docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/nci-header/cgdp-mega-menu/components/cgdp-mega-menu.tsx:76]:
```suggestion
						><span>{mmData.primary_nav_label}</span></a>
```

File: docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/templates/block/ncids/block--ncids-header.html.twig [hunk 10]
```diff
@@ -1,66 +1,64 @@
...
+              %}
+              {% if navitem.hasNcidsMegaMenu %}
+                {% set link_attribs = link_attribs.setAttribute('data-menu-id', navitem.id) %}
+              {% else %}
+                {% set link_attribs = link_attribs.setAttribute('data-megamenu-disabled', 'true') %}
+              {% endif %}
+              {% if navitem.navstate == 'ancestor' or navitem.navstate == 'current' %}
+                {% set link_attribs = link_attribs.addClass('usa-current') %}
+              {% endif %}
+              <a {{link_attribs}}>{{navitem.label}}</a>
```
Comment (olitharp-nci) [inline docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/templates/block/ncids/block--ncids-header.html.twig:56]:
```suggestion
              <a {{link_attribs}}><span>{{navitem.label}}</span></a>
```

**Review by olitharp-nci — CHANGES_REQUESTED** [review]

**Review by olitharp-nci — APPROVED** [review]

**Review by sarinapadilla — APPROVED** [review]
@bryanpizzillo @olitharp-nci , this passes product/design review per 9/7 slack thread with @nicolecrom @kate-mashkina @laurelthrash @blilianyu and myself!

**Review by arcepaul — APPROVED** [review]

**Review by bryanpizzillo — APPROVED** [review]

**Pull request merged**